app/src/main/java/org/wikipedia/captcha/CaptchaHandler.kt

```diff
@@ -0,0 +1,100 @@
+package org.wikipedia.captcha
+
+import android.app.Activity
+import android.view.View
+import androidx.appcompat.app.AppCompatActivity
+import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
+import io.reactivex.rxjava3.disposables.CompositeDisposable
+import io.reactivex.rxjava3.schedulers.Schedulers
+import org.wikipedia.R
+import org.wikipedia.databinding.GroupCaptchaBinding
+import org.wikipedia.dataclient.ServiceFactory
+import org.wikipedia.dataclient.WikiSite
+import org.wikipedia.page.LinkMovementMethodExt
+import org.wikipedia.util.DeviceUtil.hideSoftKeyboard
+import org.wikipedia.util.FeedbackUtil.showAndroidAppRequestAnAccount
+import org.wikipedia.util.FeedbackUtil.showError
+import org.wikipedia.util.StringUtil.fromHtml
+import org.wikipedia.views.ViewAnimations.crossFade
+import org.wikipedia.views.ViewUtil.loadImage
+
+class CaptchaHandler(private val activity: Activity, private val wiki: WikiSite,
+                     captchaView: View, private val primaryView: View,
+                     private val prevTitle: String, submitButtonText: String?) {
+    private val binding: GroupCaptchaBinding = GroupCaptchaBinding.bind(captchaView)
+
+    private val disposables = CompositeDisposable()
+    var token: String? = null
+    private var captchaResult: CaptchaResult? = null
+
+    fun captchaId(): String? {
+        return captchaResult?.captchaId
+    }
+
+    fun captchaWord(): String {
+        return binding.captchaText.editText?.text.toString()
+    }
+
+    fun dispose() {
+        disposables.clear()
+    }
+
+    val isActive: Boolean
+        get() = captchaResult != null
+
+    fun handleCaptcha(token: String?, captchaResult: CaptchaResult) {
+        this.token = token
+        this.captchaResult = captchaResult
+        handleCaptcha(false)
+    }
+
+    private fun handleCaptcha(isReload: Boolean) {
+        if (captchaResult == null) {
+            return
+        }
+        hideSoftKeyboard(activity)
+        if (!isReload) {
+            crossFade(primaryView, binding.captchaContainer)
+        }
+        // In case there was a captcha attempt before
+        binding.captchaText.editText?.setText("")
+        loadImage(binding.captchaImage, captchaResult!!.getCaptchaUrl(wiki), roundedCorners = false, largeRoundedSize = false, force = true, listener = null)
+    }
+
+    fun hideCaptcha() {
+        (activity as AppCompatActivity).supportActionBar!!.title = prevTitle
+        crossFade(binding.captchaContainer, primaryView)
+    }
+
+    fun cancelCaptcha() {
+        if (captchaResult == null) {
+            return
+        }
+        captchaResult = null
+        binding.captchaText.editText?.setText("")
+        hideCaptcha()
+    }
+
+    init {
```

Comment:
Can we have `init{}` to the top for consistency?

Reply:
yep